doc/PluginSystem.md

```diff
@@ -0,0 +1,152 @@
+# Plugin System and Lifecycle Management
+The Serverless Plugin System is at the core of our Framework. It supports any feature we build, it decides how fast we can build them, how easily testable they are and how approachable they are for our users.
+
+The main goals for our plugin system are:
+
+* Separation of CLI configuration and Plugin logic
+  * We want any plugin author to be able to easily create new commands within the serverless framework and extend existing commands. To achieve this we need to create a strong separation between CLI and Plugins that has clear interfaces between each other.
+* Separation between logic of different plugins
+  * Different plugins need to have an easy way to run independently and after each other without defining dependencies between each other.
+* Greater Extensibility
+  * Plugins need to be able to easily integrate into the lifecycle of a command independent of other Plugins that are running and can extend the functionality of Serverless easily.
+
+## Concepts
+
+### Plugin
+A Plugin encapsulates commands and hooks in a shareable way. A Plugin is not forced to provide both, they can only consist of a list of commands and lifecycle events or only of hooks.
+
+### Command
+A command represents a CLI command that can be called by a user, e.g. ```serverless deploy``` would be the deploy command. A command has no logic, but simply defines the CLI configuration (e.g. command, subcommands, parameters) and the lifecycle events for this particular command. Every command defines its own lifecycle events, so different commands can have completely different lifecycles. The commands that come with Serverless (e.g. deploy, remove, …) are implemented in the exact same way as Commands built by other users. This means that lifecycle events we define for our commands do not have any special meaning in Serverless or for other plugins. Every Command is free to have its own lifecycle events, none of them are more special than others.
+
+```javascript
+'use strict';
+
+class HelloWorld {
+  constructor() {
+    this.commands = {
+      deploy: {
```

Comment:
Would include the `usage` key here (this will be used by the CLI when someone e.g runs `serverless deploy --help`.